### PR TITLE
Prevent follow-up requests if a response is likely bad

### LINKED ISSUE #919: Prevent direct followup requests to a peer if a response is likely bad
### Describe the feature

currently the `StateFetcher` returns an `Ok` response back to the downloader:

https://github.com/paradigmxyz/reth/blob/115e623ae630b3c1841841dfe192a8cc63505786/crates/net/network/src/fetch/mod.rs#L216-L243

and the downloader can then report the peer if the response is bad.

at this point, the peer is still eligible for a followup request, but depending on the response we can already derive if this was a bad response, for example if 
* empty headers response, but requested more

In this case we should still send it to the downloader but not send a follow-up request since it is likely that this was a bad response.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by leruaa**
Closes #919

### CODE REVIEW

File: crates/net/network/src/fetch/mod.rs [hunk 1]
```diff
@@ -232,16 +242,31 @@ impl StateFetcher {
...
+                if peer.state.on_request_finished() {
+                    return self.followup_request(peer_id)
+                }
             }
         }
+
         None
     }
 
+    fn is_likely_bad_message(request: &HeadersRequest, headers: &[Header]) -> bool {
```
Comment (mattsse) [inline crates/net/network/src/fetch/mod.rs:258]: we will only use this on `PeerResult` so we can implement this for `PeerResult<Vec<Header>` directly, so we can call `result.is_likely_bad`

File: crates/net/network/src/fetch/mod.rs [hunk 2]
```diff
@@ -232,16 +242,31 @@ impl StateFetcher {
...
+        if request.limit != headers.len() as u64 {
+            return true
+        }
+
+        if let BlockHashOrNumber::Number(block_number) = request.start {
+            return Some(block_number) != headers.get(0).map(|h| h.number)
+        }
```
Comment (mattsse) [inline crates/net/network/src/fetch/mod.rs:265]: please convert into match statement over BlockHashOrNumber

File: crates/net/network/src/fetch/mod.rs [hunk 3]
```diff
@@ -232,16 +242,31 @@ impl StateFetcher {
...
+                    return self.followup_request(peer_id)
+                }
             }
         }
+
         None
     }
 
+    fn is_likely_bad_message(request: &HeadersRequest, headers: &[Header]) -> bool {
+        if request.limit != headers.len() as u64 {
```
Comment (mattsse) [inline crates/net/network/src/fetch/mod.rs:259]: limit is only the upper bound of the requested range, the peer may return less then requested, which is valid.

But what usually happens is that peers just return one or none headers, so we should check for `< 1  && request.limit != len`

File: crates/net/network/src/fetch/mod.rs [hunk 4]
```diff
@@ -220,7 +220,17 @@ impl StateFetcher {
...
-        if let Some(resp) = self.inflight_headers_requests.remove(&peer_id) {
+        let resp = self.inflight_headers_requests.remove(&peer_id);
+        let is_likely_a_bad_message = res
+            .as_ref()
+            .map(|headers| {
+                resp.as_ref()
+                    .map(|r| Self::is_likely_bad_message(&r.request, headers))
+                    .unwrap_or_default()
+            })
+            .unwrap_or_default();
```
Comment (mattsse) [inline crates/net/network/src/fetch/mod.rs:231]: we can get rid of this by moving into function of `PeerResult`

**Review by mattsse — CHANGES_REQUESTED** [review]
ty,
some suggestions and nits, overall this looks great.

File: crates/interfaces/src/p2p/error.rs [hunk 5]
```diff
@@ -1,14 +1,44 @@
...
+    fn is_likely_a_bad_message(&self, request: &HeadersRequest) -> bool;
+}
+
+impl RequestValidation for RequestResult<Vec<Header>> {
+    fn is_likely_a_bad_message(&self, request: &HeadersRequest) -> bool {
+        match self {
+            Ok(headers) => {
+                let request_length = headers.len() as u64;
+
+                if request_length < 1 && request.limit != request_length {
```
Comment (gakonst) [inline crates/interfaces/src/p2p/error.rs:26]:
```suggestion
                if request_length == 0 && request.limit > 0 {
```

Is this equivalent? What do we want to ensure here?

**Review by gakonst — APPROVED** [review]

**Comment by gakonst** [comment]
CI failure unrelated

File: crates/interfaces/src/p2p/error.rs (hunk 5 above)
Comment (leruaa) [reply crates/interfaces/src/p2p/error.rs:26]: Related to [this comment](https://github.com/paradigmxyz/reth/pull/946#discussion_r1083273973) above. Indeed I think your suggestion is equivalent and easier to read.

Comment (mattsse) [reply crates/interfaces/src/p2p/error.rs:26]: ah my bad, meant to type `<=1` here

bad responses are usually empty or 1 but the requested blocks range was larger

**Comment by leruaa** [comment]
@gakonst here is the  fix for CI: https://github.com/paradigmxyz/reth/pull/961

**Pull request merged**